### ISSUE TITLE
Lint to suggest clone_from and clone_into

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust-clippy#1709]
Basically, we'd suggest the following replacements:

`a = b.clone()` => `a.clone_from(&b)`
`a = b.to_owned()` => `b.clone_into(&mut a)`

**Comment by luisbg** [comment]
Is anybody working on this one? If not, I am happy to do it.

Looks like a good new contributor issue :smile:

**Comment by oli-obk** [comment]
@luisbg it's all yours! Feel free to ask questions here or open a prototype PR in any level of completeness if you have questions.

**Comment by oli-obk** [comment]
Note that this should not lint in case `a` has never been assigned before:

```rust
let a = b.clone();

/// or

let a;

....

a = b.clone();
```

**Comment by luisbg** [comment]
Great! Will start working on it and share progress soon

**Comment by llogiq** [comment]
Fantastic! Feel free to push early if you like to discuss. It's easier to discuss a WIP PR if it actually exists. :smile:

**Comment by luisbg** [comment]
I have created some boilerplate code in the following commit:
https://github.com/luisbg/rust-clippy/commit/38c59c298edf416db2c7b7eaefbb1bb7d40279d9

Wanted to be sure I am going in the right direction and not sure how to proceed with checking the expression.

I have been following @llogiq 's blogpost [0], but not clear how to do the check_expr() since lints seem to use rustc::hir now instead of syntax::ast.

Any advice? Thanks! :smiley: 


[0] https://llogiq.github.io/2015/06/04/workflows.html

**Comment by Manishearth** [comment]
`hir` is basically a different AST.

http://manishearth.github.io/rust-internals-docs/rustc/hir/enum.Expr_.html may help with figuring out how the HIR works.

**Comment by sanmai-NL** [comment]
Note that `clone_into()` hasn’t been stabilized yet.

> error: use of unstable library feature 'toowned_clone_into': recently added (see issue #412

`rustc 1.24.0-nightly (edbd7d232 2017-12-20)`

**Comment by kpreid** [comment]
This has been implemented in #12077 and can be closed. (Thanks to @Kobzol !)
